### PR TITLE
RFC: Rover motor lib

### PR DESCRIPTION
**Pull request by khancyr**
Hello,

This is a first try for a motor lib for rover !
It implement a new parameter TYPE_CLASS that will allow to separate correctly the different type of rover.
It put motor initialization and mixing in a dedicated file.
It implement a new skid steering (aka tank mode). This new mode require to change the servo function from k_throttle;k_steering to k_throttleLeft:k_throttleRight .

Other improvements:
- add some comments in various place
- correct some safety not correctly set
- always use scaled RC input and remove the need of special handling for manuals modes. It solve non-working servo/rc reverse feature
- improvement to brushed motor control
- implementation of motor test for rover
- add some pre_arm_check to prevent usage of rover without TYPE_CLASS set

It has only been tested on SITL ! 
TODO : 
- [ ] Real tests
- [ ] New autotest for new features
- [ ] implement PWM_TYPE
- [ ] implement PWM_FREQ
- [ ] impement a way to link a motor ouput with an ON/OFF pin use to switch the direction on H-Bridge
- [ ] Correct DO_REVERSE function that is broken.

Questions: 
I go on the simplier way and only add some function to ROVER class. Should we make a dedicated class for motor ? The mixing and calculus for output need some parameter from rover ...
Should I do a conversion function from SKID_STEER_OUT to TYPE_CLASS ? Currently the rover won't arm without TYPE_CLASS set. Will convert CH1 and CH3 to default of each class unless they got another function

### CODE REVIEW

**Comment by khancyr** [comment]
well motor lib should be a problem if we take throttle_slewrate and put it in motor param

**Comment by rmackay9** [comment]
@khancyr, can we break this PR up into smaller bits?  I think it is easier to get the changes included if we separate out the changes that should be able to go in immediately from those that are still WIP.

File: APMrover2/radio.cpp [hunk 1]
```diff
@@ -36,6 +29,15 @@ void Rover::set_control_channels(void)
...
 {
+    // Initialize default RC channels
+    channel_steer    = RC_Channels::rc_channel(rcmap.roll()-1);
+    channel_throttle = RC_Channels::rc_channel(rcmap.throttle()-1);
+    channel_learn    = RC_Channels::rc_channel(g.learn_channel-1);
+
+    // set rc channel ranges
+    channel_steer->set_angle(SERVO_MAX);  // -4500 to 4500 centidegrees
+    channel_throttle->set_angle(100);     // 0 - 100%
+
```
Comment (khancyr) [inline APMrover2/radio.cpp]: hummm this doesn't need to be change otherwise we can live change rcmap ...

**Comment by khancyr** [comment]
ok, the only thing to remove then is motor test function

File: libraries/AP_HAL_PX4/RCOutput.cpp [hunk 2]
```diff
@@ -329,7 +331,12 @@ void PX4RCOutput::write(uint8_t ch, uint16_t period_us)
         _max_channel = ch + 1;
     }
 
+
```
Comment (rmackay9) [inline libraries/AP_HAL_PX4/RCOutput.cpp]: There's one unnecessary line added.. probably a mistake so I'd remove it although obviously not a big deal.

File: APMrover2/motors.cpp [hunk 3]
```diff
@@ -76,6 +76,10 @@ void Rover::output_to_motors()
         // limit throttle movement speed
         if (g.throttle_slewrate > 0) {
             SRV_Channels::limit_slew_rate(SRV_Channel::k_throttle, g.throttle_slewrate, G_Dt);
+            if (isTypeTank) {
+                // when skid steering also limit 2nd channel
+                SRV_Channels::limit_slew_rate(SRV_Channel::k_steering, g.throttle_slewrate, G_Dt);
```
Comment (khancyr) [inline APMrover2/motors.cpp]: humm that doesn't seems good ...

Comment (khancyr) [inline APMrover2/motors.cpp]: well it works on pure skid steering
If you got servo that should use k_steering you are limiting them ...

Comment (khancyr) [inline APMrover2/motors.cpp]: well I need to find a velocity/mecanical model to the calculus

**Comment by khancyr** [comment]
close for improvement

**Pull request closed without merging**